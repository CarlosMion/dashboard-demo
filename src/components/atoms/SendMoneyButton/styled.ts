"use client";

import { styled } from "@mui/material/styles";
import Button from "@mui/material/Button";
import Stack from "@mui/material/Stack";

export const ButtonContainer = styled(Button)(({ theme }) => ({
  display: "flex",
  flexDirection: "row",
  padding: `${theme.spacing(1.8)} ${theme.spacing(3)}`,
  backgroundColor: theme.palette.text.primary,
  borderRadius: "50px",
  textTransform: "none",
  color: theme.palette.common.white,
  fontWeight: 400,
  fontSize: "0.75rem",
}));

export const Content = styled(Stack)(({ theme }) => ({
  flexDirection: "row",
  gap: theme.spacing(1.5),
  alignItems: "center",
  justifyContent: "center",
}));
