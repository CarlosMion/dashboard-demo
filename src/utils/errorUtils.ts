export type ServerError = {
  error: string;
  statusCode?: number;
};

export const isServerErrorWithStatusCode = (
  result: object
): result is ServerError => {
  return "error" in result && "statusCode" in result;
};

export const isServerError = (result: unknown): result is ServerError => {
  return typeof result === "object" && result !== null && "error" in result;
};

type ErrorWithMessage = {
  message: string;
};

export const isErrorWithMessage = (
  error: unknown
): error is ErrorWithMessage => {
  return (
    typeof error === "object" &&
    error !== null &&
    "message" in error &&
    typeof (error as Record<string, unknown>).message === "string"
  );
};

export const toErrorWithMessage = (maybeError: unknown): ErrorWithMessage => {
  if (isErrorWithMessage(maybeError)) return maybeError;

  try {
    return new Error(JSON.stringify(maybeError));
  } catch {
    // fallback in case there's an error stringifying the maybeError
    // like with circular references for example.
    return new Error(String(maybeError));
  }
};

export const getErrorMessage = (error: unknown): string => {
  return toErrorWithMessage(error).message;
};

export class ApiError extends Error {
  statusCode: number;

  constructor(message: string, statusCode: number) {
    super(message);
    this.name = "ApiError";
    this.statusCode = statusCode || 500;
  }
}
