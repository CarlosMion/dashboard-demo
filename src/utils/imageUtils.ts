export const dataURItoBlob = (dataURI: string, fileType: string): Blob => {
  const byteString = atob(dataURI.split(',')[1]);
  const arrayBuffer = new ArrayBuffer(byteString.length);
  const intArray = new Uint8Array(arrayBuffer);

  for (let i = 0; i < byteString.length; i++) {
    intArray[i] = byteString.charCodeAt(i);
  }

  return new Blob([arrayBuffer], { type: fileType });
};

export const dataURItoFile = (
  dataURI: string,
  fileName: string,
  fileType: string
): File => {
  const blob = dataURItoBlob(dataURI, fileType);
  return new File([blob], fileName, { type: blob.type });
};

export const blobToFile = (blob: Blob, fileName: string): File => {
  return new File([blob], fileName, { type: blob.type });
};

export const getImageDimensions = (
  src: string
): Promise<{ width: number; height: number } | null> => {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.src = src;

    img.onload = () =>
      resolve(
        img.width && img.height
          ? { width: img.width, height: img.height }
          : null
      );
    img.onerror = (error) => reject(error);
  });
};

export const isImageDimensionWithinThreshold = (
  actual: number,
  ideal: number,
  threshold: number
): boolean => {
  const isWithinThreshold = actual >= ideal * (1 - threshold);
  return isWithinThreshold;
};
