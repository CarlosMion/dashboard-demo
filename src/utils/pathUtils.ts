export function toCamelCase(str?: string): string | undefined {
  return str?.replace(/-([a-z])/g, (match, letter) => letter.toUpperCase());
}
