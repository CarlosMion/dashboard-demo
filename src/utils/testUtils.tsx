import { render } from "@testing-library/react";
import { ReactNode } from "react";
import AppProviders from "@/providers/AppProviders/AppProviders";

import messages from "@/i18n/messages/en.json";

const customRender = (children: ReactNode) =>
  render(
    <AppProviders
      locale={"en"}
      messages={messages}
      userAgent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
    >
      {children}
    </AppProviders>
  );

export * from "@testing-library/react";

export { customRender as render };
