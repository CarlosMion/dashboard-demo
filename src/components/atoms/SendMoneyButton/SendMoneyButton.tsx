import { ButtonContainer, Content } from "./styled";
import { useTranslations } from "next-intl";
import { SendIcon } from "@/components/atoms/icons";
import { useCallback } from "react";
import { toast } from "react-toastify";

export default function SendMoneyButton() {
  const t = useTranslations();

  const sentToast = useCallback(
    () => toast.success(t("moneySentSuccessfully")),
    [t]
  );

  return (
    <ButtonContainer onClick={sentToast}>
      <Content>
        {t("send")}
        <SendIcon />
      </Content>
    </ButtonContainer>
  );
}
