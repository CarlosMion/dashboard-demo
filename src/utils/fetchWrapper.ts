import { isServerError } from "./errorUtils";

const fetcher = async <T>(
  info: RequestInfo | URL,
  init?: RequestInit
): Promise<T> => {
  let result;
  try {
    const injectedHeaders: Record<string, string> = {};
    if (typeof window === "undefined") {
      const headers = await import("next/headers");
      (await headers.headers()).forEach((value, key) => {
        injectedHeaders[key] = value;
      });
    }

    const buildHeaders: Record<string, string> = {};
    let url = info;
    if (typeof info === "string") {
      url = `${process.env.NEXT_PUBLIC_WEBAPP_API_URL}/${info}`;
    }
    const reqHeaders = {
      "Content-Type": "application/json",
      ...init?.headers,
      ...injectedHeaders,
      ...buildHeaders,
    };
    result = await fetch(url, {
      ...init,
      headers: reqHeaders,
    });
    if (!result.ok) {
      throw new Error("Failed to fetch");
    }
    let parsedResult;
    try {
      parsedResult = await result.json();
    } catch (error) {
      console.log("Failed to parse response, probably not JSON", result, error);
      return result as T;
    }
    if (isServerError(parsedResult)) {
      throw new Error(parsedResult.error);
    }
    return parsedResult as T;
  } catch (error) {
    throw error;
  }
};

export default fetcher;
