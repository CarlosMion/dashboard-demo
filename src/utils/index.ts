import { isEdge } from "./userAgentUtils";
import { maskCardNumber } from "./stringUtils";
import { toCamelCase } from "./pathUtils";
import fetcher from "./fetchWrapper";
import {
  isServerErrorWithStatusCode,
  ApiError,
  getErrorMessage,
  isErrorWithMessage,
  isServerError,
  toErrorWithMessage,
} from "./errorUtils";
import {
  blobToFile,
  dataURItoBlob,
  dataURItoFile,
  getImageDimensions,
  isImageDimensionWithinThreshold,
} from "./imageUtils";
import {
  createImage,
  getRadianAngle,
  rotateSize,
  getCroppedImg,
} from "./cropImageUtils";
export {
  fetcher,
  isEdge,
  toCamelCase,
  isServerErrorWithStatusCode,
  ApiError,
  getErrorMessage,
  isErrorWithMessage,
  isServerError,
  toErrorWithMessage,
  blobToFile,
  dataURItoBlob,
  dataURItoFile,
  getImageDimensions,
  isImageDimensionWithinThreshold,
  getCroppedImg,
  rotateSize,
  getRadianAngle,
  createImage,
  maskCardNumber,
};
