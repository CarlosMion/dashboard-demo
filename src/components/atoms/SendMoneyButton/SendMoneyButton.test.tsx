import { render, screen, fireEvent } from "@/utils/testUtils";
import SendMoneyButton from "./SendMoneyButton";
import { toast } from "react-toastify";

// Mock react-toastify
jest.mock("react-toastify", () => ({
  toast: {
    success: jest.fn(),
  },
}));

const toastMock = jest.spyOn(toast, "success");

describe("SendMoneyButton", () => {
  it("renders correctly and matches snapshot", () => {
    const { container } = render(<SendMoneyButton />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("displays the correct translated text", () => {
    render(<SendMoneyButton />);
    expect(screen.getByText("Send")).toBeInTheDocument();
  });

  it("renders the SendIcon", () => {
    render(<SendMoneyButton />);
    expect(screen.getByRole("img", { hidden: true })).toBeInTheDocument();
  });

  it("shows success toast when clicked", () => {
    render(<SendMoneyButton />);
    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(toast.success).toHaveBeenCalledWith("Money sent successfully");
  });

  it("shows success toast only once per click", () => {
    render(<SendMoneyButton />);
    const button = screen.getByRole("button");
    jest.clearAllMocks();
    fireEvent.click(button);
    fireEvent.click(button);
    expect(toastMock).toHaveBeenCalledTimes(2);
  });
});
