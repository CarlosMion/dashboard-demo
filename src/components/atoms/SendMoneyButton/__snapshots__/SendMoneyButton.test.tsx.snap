// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendMoneyButton renders correctly and matches snapshot 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-117n223-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  <div
    class="MuiStack-root css-5l7in0-MuiStack-root"
  >
    Send
    <svg
      aria-busy="false"
      aria-hidden="true"
      class="MuiBox-root css-7153nf"
      fill="none"
      role="img"
      viewBox="0 0 21 18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20.9857 0.726043C21.0881 0.26211 20.6209 -0.129073 20.1652 0.0402252L0.395798 7.38713C0.157869 7.47558 0.000493341 7.69758 1.15777e-06 7.9455C-0.000491025 8.19346 0.155982 8.41606 0.393584 8.50538L5.94717 10.5935V17.1703C5.94717 17.4477 6.14286 17.6888 6.42012 17.7529C6.69554 17.8167 6.98273 17.6891 7.11184 17.4398L9.40874 13.0026L15.014 17.0523C15.355 17.2986 15.8451 17.1419 15.9692 16.7461C21.2027 0.0415434 20.9764 0.768008 20.9857 0.726043V0.726043ZM16.1057 2.83101L6.4752 9.50787L2.33151 7.94993L16.1057 2.83101ZM7.17763 10.4914L15.5721 4.67153C8.34878 12.0899 8.72604 11.6994 8.69454 11.7407C8.64774 11.802 8.77595 11.5631 7.17763 14.6508V10.4914ZM15.0467 15.5842L10.1128 12.0195L19.034 2.85737L15.0467 15.5842Z"
        fill="#FFFFFF"
      />
    </svg>
  </div>
</button>
`;
